### PR TITLE
[incubator/openmpi] generic openmpi support

### PR DESCRIPTION
**Pull request by cheyang**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

This PR is for running Openmpi easily with helm by generating machine list file automatically.

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by timstoop** [comment]
I don't mind taking a look at it this evening, when I have a bit more time, but I overlook a lot of things with such large commits, so I tend to leave them to the more experienced reviewers. It's not good for my ego when I add a LGTM and the next reviewer finds several glaring problems ;-)

**Comment by cheyang** [comment]
I see, thanks. @timstoop

File: incubator/openmpi/README.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+|-----------|-------------|---------|
+| `mpiWorker.number`|  The mpi worker's number | `5` |
+| `mpiWorker.image.repository` | image repository for the service openmpi | `cheyang/openmpi:3.0.0` |
+| `mpiWorker.image.tag` | image repository for the service openmpi | `cheyang/openmpi:3.0.0` |
+| `mpiWorker.image.pullPolicy` | `imagePullPolicy` for the service openmpi | `IfNotPresent` |
+| `mpiWorker.sshPort` | mpiWorker's sshPort | `22` |
+| `mpiWorker.podManagementPolicy` | mpiWorker's podManagementPolicy | `Parallel` |
+| `mpiWorker.env` | mpiWorker's environment varaibles | `{}` |
+| `persistence.pvc.storage`| the storage size to request | `5Gi` |
+| `persistence.pvc.matchLabels`| the selector for pv | `{}` |
```
Comment (timstoop) [inline incubator/openmpi/README.md]: I know this is nit-picky, but can you please align the pipe-symbols?

File: incubator/openmpi/values.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+mpiWorker:
+  number: 5
+  podManagementPolicy: Parallel
+  image:
+    repository: cheyang/openmpi
+    tag: 3.0.0
+    pullPolicy: IfNotPresent
+  sshPort: 22
+  resources: {}
+  privileged: true
```
Comment (timstoop) [inline incubator/openmpi/values.yaml]: This line needs to move either up or down, so `resources` is 'connected' to the example resources.

**Review by timstoop — CHANGES_REQUESTED** [review]
Clear and concise, I like it.

**Comment by cheyang** [comment]
Updated based on comments, thx.

File: incubator/openmpi/templates/config.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+{{- $openmpiName := include "openmpi.fullname" . }}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ template "openmpi.fullname" . }}-cm
+  labels:
+    heritage: {{ .Release.Service | quote }}
+    release: {{ .Release.Name | quote }}
+    chart: {{ template "openmpi.chart" . }}
+    app: {{ template "openmpi.fullname" . }}
```
Comment (unguiculus) [inline incubator/openmpi/templates/config.yaml:11]: app: {{ template "openmpi.name" . }}

File: incubator/openmpi/templates/statefulset.yaml [hunk 4]
```diff
@@ -0,0 +1,75 @@
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: {{ template "openmpi.fullname" . }}
+  labels:
+    app: {{ template "openmpi.name" . }}
+    chart: {{ template "openmpi.chart" . }}
```
Comment (unguiculus) [inline incubator/openmpi/templates/statefulset.yaml:7]: Remove `release` and `heritage` labels from selector.

File: incubator/openmpi/templates/statefulset.yaml [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+    role: mpiworker
+spec:
+  selector:
+    matchLabels:
+      app: {{ template "openmpi.name" . }}
+      chart: {{ template "openmpi.chart" . }}
```
Comment (unguiculus) [inline incubator/openmpi/templates/statefulset.yaml:15]: Remove `release` and `heritage` labels from selector.

File: incubator/openmpi/templates/config.yaml [hunk 6]
```diff
@@ -0,0 +1,20 @@
+{{- $workerNum := .Values.mpiWorker.number -}}
+{{- $openmpiName := include "openmpi.fullname" . }}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ template "openmpi.fullname" . }}-cm
```
Comment (unguiculus) [inline incubator/openmpi/templates/config.yaml:6]: I'd remove the suffix from the name.

File: incubator/openmpi/templates/service.yaml [hunk 7]
```diff
@@ -0,0 +1,18 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ template "openmpi.fullname" . }}-svc
```
Comment (unguiculus) [inline incubator/openmpi/templates/service.yaml:4]: I'd remove the suffix from the name.

File: incubator/openmpi/templates/statefulset.yaml [hunk 8]
```diff
@@ -0,0 +1,75 @@
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: {{ template "openmpi.fullname" . }}
+  labels:
+    app: {{ template "openmpi.name" . }}
+    chart: {{ template "openmpi.chart" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+    role: mpiworker
```
Comment (unguiculus) [inline incubator/openmpi/templates/statefulset.yaml:10]: Is the `role` label required, since this is the only role I see here?

**Review by unguiculus — CHANGES_REQUESTED** [review]

**Pull request closed without merging**